android-core/src/main/kotlin/com/mparticle/internal/ApplicationContextWrapper.kt

```diff
@@ -0,0 +1,351 @@
+            recordedLifecycleList.add(LifeCycleEvent(MethodType.ON_PAUSED, WeakReference(activity)))
+        }
+
+        override fun onActivityStopped(activity: Activity) {
+            recordedLifecycleList.add(
+                LifeCycleEvent(
+                    MethodType.ON_STOPPED,
+                    WeakReference(activity)
+                )
+            )
+        }
+
+        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
+            recordedLifecycleList.add(
+                LifeCycleEvent(
+                    MethodType.ON_SAVE_INSTANCE_STATE,
+                    WeakReference(activity),
+                    outState
+                )
+            )
+        }
+
+        override fun onActivityDestroyed(activity: Activity) {
+            recordedLifecycleList.add(
+                LifeCycleEvent(
+                    MethodType.ON_DESTROYED,
+                    WeakReference(activity)
+                )
+            )
+        }
+
+        private val recordedLifecycleList: MutableList<LifeCycleEvent>
+            get() {
+                if (lifeCycleEvents.size > MAX_LIST_SIZE) {
+                    lifeCycleEvents.removeAt(0)
+                    return recordedLifecycleList
+                }
+                return lifeCycleEvents
+            }
+
+        internal val recordedLifecycleListCopy: LinkedList<LifeCycleEvent>
+            get() {
+                var list: LinkedList<LifeCycleEvent>
+                synchronized(lifeCycleEvents) {
+                    list = LinkedList(lifeCycleEvents)
+                }
+                return list
+            }
+    }
+
+    inner class LifeCycleEvent(
+        val methodType: MethodType,
+        val activityRef: WeakReference<Activity>?,
+        val bundle: Bundle?
+    ) {
+        constructor(
+            methodType: MethodType,
+            activityRef: WeakReference<Activity>?
+        ) : this(methodType, activityRef, null)
+
+        override fun equals(o: Any?): Boolean {
+            if (o is LifeCycleEvent) {
+                val l = o
+                var matchingActivityRef = false
+                if (l.activityRef == null && activityRef == null) {
+                    matchingActivityRef = true
+                } else if (l.activityRef != null && activityRef != null) {
+                    matchingActivityRef = l.activityRef.get() === activityRef.get()
+                }
+                return matchingActivityRef && l.methodType == methodType && l.bundle == bundle
+            }
+            return false
+        }
+    }
+
+    internal inner class ReplayLifecycleCallbacksRunnable(var callback: ActivityLifecycleCallbacks) :
+        Runnable {
+        @TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH)
+        override fun run() {
+            if (callback != null && mActivityLifecycleCallbackRecorder != null && isReplayActivityLifecycle) {
+                val reference = if (MParticle.getInstance()!!
```

Comment:
I'm seeing the non-null assertion operator used all over the place here...

We can not use this in production code ever, unless it's absolutely necessary for some reason and immediately preceeded by a null check or the variable declaration or assignment of a non-null value (and even then we should not use it except as a last resort).

Please see the Kotlin docs here: https://kotlinlang.org/docs/null-safety.html#not-null-assertion-operator